[mlir][OpenMP] Pack task private variables into a heap-allocated context struct

tblah opened the pull request:
See RFC:
https://discourse.llvm.org/t/rfc-openmp-supporting-delayed-task-execution-with-firstprivate-variables/83084

The aim here is to ensure that tasks which are not executed for a while after they are created do not try to reference any data which are now out of scope. This is done by packing the data referred to by the task into a heap allocated structure (freed at the end of the task).

I decided to create the task context structure in
OpenMPToLLVMIRTranslation instead of adapting how it is done CodeExtractor (via OpenMPIRBuilder] because CodeExtractor is (at least in theory) generic code which could have other unrelated uses.

tblah committed:
[mlir][OpenMP] Pack task private variables into a heap-allocated context struct

See RFC:
https://discourse.llvm.org/t/rfc-openmp-supporting-delayed-task-execution-with-firstprivate-variables/83084

The aim here is to ensure that tasks which are not executed for a while
after they are created do not try to reference any data which are now
out of scope. This is done by packing the data referred to by the task
into a heap allocated structure (freed at the end of the task).

I decided to create the task context structure in
OpenMPToLLVMIRTranslation instead of adapting how it is done
CodeExtractor (via OpenMPIRBuilder] because CodeExtractor is (at least
in theory) generic code which could have other unrelated uses.

tblah commented:
Pull requests in this stack:

1. https://github.com/llvm/llvm-project/pull/125303
2. https://github.com/llvm/llvm-project/pull/125304
3. https://github.com/llvm/llvm-project/pull/125307

tblah committed:
Allocate simple private vars inside of the task

There's no need to put the variable in the context structure if it can
be initialized without reading from the mold variable.

This leads to much simpler code generation for nested tasks which only
privatize simple scalar variables and do not use firstprivate. This
works around the failure in 0226_0013 (Fujitsu test).

tblah commented:
This should now be ready for review.

ergawy reviewed (commented):
Thanks Tom! I have a few comments. Not much of an expert when it comes to tasks so I might have missed something worth noting.

tblah committed:
Add methods to operation to test if the mold arg is read

tblah committed:
Fix comment

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1730,6 +1730,126 @@ buildDependData(std::optional<ArrayAttr> dependKinds, OperandRange dependVars,
+private:
+  llvm::IRBuilderBase &builder;
+  LLVM::ModuleTranslation &moduleTranslation;
+  MutableArrayRef<omp::PrivateClauseOp> privateDecls;
```
I agree it should be, but unfortunately the table-gen accessors (e.g. `getRegions()`, `getCopyRegion()`) for the mlir operation aren't marked `const`. This makes sense because the region can be mutated through this reference, but it is a shame there aren't any `const` alternatives.

The non-const regions means that `getInitMoldArg()` etc cannot be `const`.

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1730,6 +1730,126 @@ buildDependData(std::optional<ArrayAttr> dependKinds, OperandRange dependVars,
+  /// privateVarTypes, privateDecls, and the elements of the structure should
+  /// all have the same order (although privateDecls which do not read from the
+  /// mold argument are skipped).
+  void generateTaskContextStruct();
```
Would you mind me leaving it as it is? Personally I find it easier to understand with it being more explicit what happens when. Especially because this does create operations.

ergawy replied:
Nope, I don't mind.

skatrak replied:
Related to @ergawy's comment, wouldn't it make sense to have this class follow a RAII pattern? So, it would call `generateTaskContextStruct()` when constructed and `freeStructPtr()` when destructed. It appears that would work well with how it's intended to be used.

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1796,36 +1918,110 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
       return handleError(std::move(err), *taskOp.getOperation());
+
+    llvm::IRBuilderBase::InsertPointGuard guard(builder);
+    builder.SetInsertPoint(builder.GetInsertBlock()->getTerminator());
```
Yes. I create all of the block structure before doing anything else so that it is in a known state.

The builder's insertion point is set here https://github.com/llvm/llvm-project/blob/d6d50cd96025d1aa74cbdb7f5da604cf104d9afc/mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp#L1920

Then `initPrivateVar` might create one or more blocks for the init region but these will always be terminated.

This line is needed exactly because the block is terminated. If new init blocks were created in `initPrivateVar`, the builder will currently have an insertion point pointing right after the terminator. We want to add any new instructions for the hack for character boxes before the terminator, which is what this line does (I don't like the naming of `SetInsertPoint` in this case: it should be `SetInsertPointBefore`).

tblah committed:
Make logic clearer

tblah committed:
Move llvm private vars into task context struct manager

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1796,36 +1918,110 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
-    llvm::Value *llvmPrivateVar = builder.CreateAlloca(
-        llvmAllocType, /*ArraySize=*/nullptr, "omp.private.alloc");
+  // Create task variable structure
+  llvm::SmallVector<llvm::Value *> privateVarAllocations;
```
I did this in [dd07fa7](https://github.com/llvm/llvm-project/pull/125307/commits/dd07fa7ae85c3837862b7771d2ad5ed449e87f16), but I'm unsure because it doesn't provide any additional encapsulation because write access is needed anyway for the charbox workaround after initializing the variables.

bhandarkar-pranav reviewed (commented):
Thank you for the PR, Tom. I have some minor comments that may amount to nitpicking.

tblah committed:
Fix comment

tblah committed:
Reserve the correct size

tblah committed:
Update charbox comment

tblah committed:
Rename llvmPrivateVars to llvmPrivateVarGEPs

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1794,37 +1909,118 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
-                       blockArg, llvmPrivateVar, llvmPrivateVars, initBlock);
-    if (err)
+                       blockArg, llvmPrivateVarAlloc, initBlock);
+    if (auto err = privateVarOrErr.takeError())
```
Same comment as above, I'm not sure if this handling can cause issues. I'd suggest:
```c++
if (!privateVarOrErr)
  return handleError(privateVarOrErr.takeError(), *taskOp.getOperation());
```

tblah replied:
Thanks. Fixed in [459d2d5](https://github.com/llvm/llvm-project/pull/125307/commits/459d2d5ea7f803cc6553b7d9007283281263f1fe)

This is a bit hard to find on the diff after the rebase unfortunately.

skatrak replied:
Seems like this instance somehow remained unchanged.

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1794,37 +1909,118 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
+       llvm::zip_equal(privateDecls, mlirPrivateVars, privateBlockArgs,
+                       taskStructMgr.getLLVMPrivateVarGEPs())) {
+    if (!privDecl.readsFromMold())
+      // to be handled inside the task
```
Nit: Move comment above 'if'.
```suggestion
      // To be handled inside the task.
```

skatrak replied:
It looks like this stayed unchanged as well.

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1794,37 +1909,118 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
+    for (auto [blockArg, privDecl, mlirPrivVar] :
+         llvm::zip_equal(privateBlockArgs, privateDecls, mlirPrivateVars)) {
+      if (privDecl.readsFromMold())
+        // This is handled before the task executes
```
Nit: Move comment above 'if'.

tblah replied:
Thanks. Fixed in [459d2d5](https://github.com/llvm/llvm-project/pull/125307/commits/459d2d5ea7f803cc6553b7d9007283281263f1fe)

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1794,37 +1909,118 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
+        // This is handled before the task executes
+        continue;
+
+      auto codegenInsertionPt = builder.saveIP();
```
Nit: This seems to be replaceable with an insertion guard.

tblah replied:
Thanks. Fixed in [459d2d5](https://github.com/llvm/llvm-project/pull/125307/commits/459d2d5ea7f803cc6553b7d9007283281263f1fe)

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1794,37 +1909,118 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
+      llvm::Expected<llvm::Value *> privateVarOrError =
+          initPrivateVar(builder, moduleTranslation, privDecl, mlirPrivVar,
+                         blockArg, llvmPrivateVar, privInitBlock);
+      if (auto err = privateVarOrError.takeError())
+        return err;
```
```suggestion
      if (!privateVarOrError)
        return privateVarOrError.takeError();
```

tblah replied:
Thanks. Fixed in [459d2d5](https://github.com/llvm/llvm-project/pull/125307/commits/459d2d5ea7f803cc6553b7d9007283281263f1fe)

skatrak replied:
Same here, it appears to have remained unchanged.

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1794,37 +1909,118 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
+    for (auto [blockArg, llvmPrivateVar] :
+         llvm::zip_equal(privateBlockArgs, llvmPrivateVars)) {
+      if (!llvmPrivateVar)
+        // This was handled above
```
Nit: Move comment above 'if'.
```suggestion
        // This was handled above.
```

tblah replied:
Thanks. Fixed in [459d2d5](https://github.com/llvm/llvm-project/pull/125307/commits/459d2d5ea7f803cc6553b7d9007283281263f1fe)

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1794,37 +1909,118 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
+    // structure
+    taskStructMgr.createGEPsToPrivateVars();
+    llvm::copy(taskStructMgr.getLLVMPrivateVarGEPs(),
+               std::back_inserter(llvmPrivateVars));
```
Could this cause `null` pointer dereferences on other uses of `llvmPrivateVars` after the new loop introduced below? For example, that list is passed to `cleanupPrivateVars`, which doesn't seem to check for `null` elements.

tblah replied:
Great point. Yeah I think it could have done. I have fixed this in https://github.com/llvm/llvm-project/pull/125307/commits/459d2d5ea7f803cc6553b7d9007283281263f1fe

skatrak reviewed (commented):
Thank you Tom, I have some minor comments but otherwise this LGTM.

tblah committed:
Fix skatrak's comments

tblah commented on mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:
```diff
@@ -146,13 +146,24 @@ def PrivateClauseOp : OpenMP_Op<"private", [IsolatedFromAbove, RecipeInterface]>
+    /// Returns true if the init region might read from the mold argument
+    bool initReadsFromMold() {
+      BlockArgument moldArg = getInitMoldArg();
+      return moldArg ? !moldArg.use_empty() : false;
```
Thanks. Fixed in [459d2d5](https://github.com/llvm/llvm-project/pull/125307/commits/459d2d5ea7f803cc6553b7d9007283281263f1fe)

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1415,11 +1413,13 @@ static llvm::Expected<llvm::BasicBlock *> allocateAndInitPrivateVars(
-        llvmPrivateVar, llvmPrivateVars, privInitBlock, mappedPrivateVars);
-    if (err)
+        llvmPrivateVar, privInitBlock, mappedPrivateVars);
+    if (auto err = privateVarOrError.takeError())
```
Thanks. Fixed in [459d2d5](https://github.com/llvm/llvm-project/pull/125307/commits/459d2d5ea7f803cc6553b7d9007283281263f1fe)

tblah commented:
This is now rebased on top of current main and ready for review. Apologies for the delay, I was on holiday and rebasing required rethinking how insertion points worked.

@ergawy I changed the signature of `initPrivateVar` so that the correct private variable SSA value is returned rather than written to an iterator argument, because this felt like an easier to understand interface.

ergawy reviewed (approved):
LGTM! Thanks!

skatrak reviewed (approved):
It appears a couple of the changes got somehow lost, but no need for another review from me. LGTM, thanks!

tblah committed:
Fix comments I missed

tblah commented:
Apologies for the comments I missed. I'm not sure how that happened. Thanks for carefully checking each one.
